pkg/kubelet/cm/dra/manager.go

```diff
@@ -381,13 +351,17 @@ func (m *ManagerImpl) GetResources(pod *v1.Pod, container *v1.Container) (*Conta
 					return fmt.Errorf("unable to get claim info for claim %s in namespace %s", *claimName, pod.Namespace)
 				}
 
-				claimAnnotations := claimInfo.annotationsAsList()
-				klog.V(3).InfoS("Add resource annotations", "claim", *claimName, "annotations", claimAnnotations)
-				annotations = append(annotations, claimAnnotations...)
-
-				devices := claimInfo.cdiDevicesAsList()
-				klog.V(3).InfoS("Add CDI devices", "claim", *claimName, "CDI devices", devices)
-				cdiDevices = append(cdiDevices, devices...)
+				// As of Kubernetes 1.31, CDI device IDs are not passed via annotations anymore.
+				for _, driverData := range claimInfo.Drivers {
+					for requestName, devices := range driverData.CDIDevices {
+						claimRequestName := claim.Request
+						if claimRequestName == "" || claimRequestName == requestName {
+							for _, deviceID := range devices {
+								cdiDevices = append(cdiDevices, kubecontainer.CDIDevice{Name: deviceID})
+							}
+						}
+					}
+				}
```

Comment:
We will need to do more than this. We need to be able to map a set of allocated devices to the driver it was allocated from. It is fine to change the `ContainerInfo` type to support this. Maybe we just structure it similar to your `DriverState` object.

Reply:
> We need to be able to map a set of allocated devices to the driver it was allocated from.

Where do we need that? Is there some missing code where something should be done that currently isn't happening?

Reply:
External clients that consume the PodResources API need to be able to map a set of devices to the driver that allocated them. Previously this could be done through the class (which had a 1-1 mapping to a driver name). Now we need to do this some other way.

Reply:
> External clients that consume the PodResources API need to be able to map a set of devices to the driver that allocated them.

But isn't it reporting on claims, not devices? There is no 1:1 relationship between claim and device.

> Previously this could be done through the class (which had a 1-1 mapping to a driver name). 

Not if the claim contained allocated resources from multiple different drivers.

We need to have a close look at what that PodResource API says about DRA...

Reply:
What kubelet knows:
- prepared claims
- which pods use a claim
- allocated devices per claim (ID = driver name/pool name/device name)
- which DRA driver handles a device (it's part of the ID)

This is what we can commit to provide to PodResource API consumers with reasonable confidence.

Reply:
Let's update this based on your proposal from this comment above:
https://github.com/kubernetes/kubernetes/pull/125488/files/171038cdaeb6311bcc5f67acf8f904e155db32b1..93ab1a4e8a5f6d9487613d08fef63b1efb0ec161#r1675607070